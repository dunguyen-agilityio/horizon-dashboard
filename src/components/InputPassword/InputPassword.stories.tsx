import type { Meta, StoryObj } from '@storybook/react';

import InputPassword from '@/components/InputPassword';
import { fn } from '@storybook/test';

const meta = {
  title: 'Components/InputPassword',
  component: InputPassword,
  tags: ['autodocs'],
} satisfies Meta<typeof InputPassword>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Default: Story = {
  args: { onChange: fn() },
};

export const InputPasswordWithDefaultValue: Story = {
  args: {
    ...Default.args,
    defaultValue: 'This is the default',
  },
};

export const InputPasswordIsDisable: Story = {
  args: {
    ...InputPasswordWithDefaultValue.args,
    isDisabled: true,
  },
};

export const InputPasswordWithErrorMessage: Story = {
  args: {
    ...InputPasswordWithDefaultValue.args,
    errorMessage: 'Password is not correct',
    isInvalid: true,
  },
};

export const InputPasswordIsReadOnly: Story = {
  args: {
    onChange: fn(),
    value: 'HandlePasswordWithLove@123',
    isReadOnly: true,
  },
};

export const InputPasswordWithPlaceholder: Story = {
  args: {
    onChange: fn(),
    placeholder: 'Password is read only',
  },
};
