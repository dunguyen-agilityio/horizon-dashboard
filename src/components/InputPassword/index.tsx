'use client';

import { useState } from 'react';

// Components
import BoxIcon from '@/components/BoxIcon';
import { Input } from '@nextui-org/input';

// Icons
import { EyeFilled, EyeSlashFilled } from '@/icons';

interface IInputPasswordProps {
  isReadOnly?: boolean;
  isDisabled?: boolean;
  isInvalid?: boolean;
  value?: string;
  defaultValue?: string;
  placeholder?: string;
  errorMessage?: string;
  onChange: (value: string) => void;
}

const InputPassword = ({
  defaultValue,
  errorMessage = '',
  isDisabled = false,
  isInvalid = false,
  isReadOnly = false,
  placeholder = 'Min. 8 characters',
  value,
  onChange,
}: IInputPasswordProps) => {
  const [isVisible, setIsVisible] = useState(false);

  const toggleVisibility = () => setIsVisible(!isVisible);

  const handleOnChange = (value: string) => onChange(value);

  return (
    <Input
      data-testid="input-password"
      aria-label="inputPassword"
      isRequired
      isInvalid={isInvalid}
      isDisabled={isDisabled}
      isReadOnly={isReadOnly}
      size="lg"
      key="password"
      label="Password"
      type={isVisible ? 'text' : 'password'}
      labelPlacement="outside"
      placeholder={placeholder}
      value={value}
      defaultValue={defaultValue}
      errorMessage={errorMessage}
      onChange={(e) => handleOnChange(e.target.value)}
      endContent={
        <button
          className="focus:outline-none"
          type="button"
          onClick={toggleVisibility}
          aria-label="toggle password visibility"
        >
          {isVisible ? (
            <BoxIcon icon={<EyeFilled />} customClass="fill-secondary" />
          ) : (
            <BoxIcon icon={<EyeSlashFilled />} customClass="fill-secondary" />
          )}
        </button>
      }
    />
  );
};

export default InputPassword;
