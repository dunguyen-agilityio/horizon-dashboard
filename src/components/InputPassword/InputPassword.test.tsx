import { act, fireEvent, render, waitFor } from '@testing-library/react';

// Components
import InputPassword from '@/components/InputPassword';

const inputPlaceholder = 'This is default placeholder';

describe('InputPassword tests', () => {
  let onChangeMock: jest.Mock;

  beforeEach(() => {
    onChangeMock = jest.fn();
  });

  it('Should match snapshot', () => {
    const { container } = render(<InputPassword onChange={onChangeMock} />);
    expect(container).toMatchSnapshot();
  });

  it('Should render EyeSlashFilled icon', () => {
    const { getByTestId } = render(<InputPassword onChange={onChangeMock} />);
    expect(getByTestId('eye-slash-filled-icon')).toBeInTheDocument();
  });

  it('Should render EyeFilled icon when click EyeSlashFilled', () => {
    const { getByTestId } = render(<InputPassword onChange={onChangeMock} />);
    const eyeSlashFilledIcon = getByTestId('eye-slash-filled-icon');

    fireEvent.click(eyeSlashFilledIcon);
    expect(getByTestId('eye-filled-icon')).toBeInTheDocument();
  });

  it('Should render Password label', () => {
    const { getByLabelText } = render(
      <InputPassword onChange={onChangeMock} />,
    );
    expect(getByLabelText('Password')).toBeInTheDocument();
  });

  it('Should render placeholder Password input', () => {
    const { getByPlaceholderText } = render(
      <InputPassword placeholder={inputPlaceholder} onChange={onChangeMock} />,
    );
    expect(getByPlaceholderText(inputPlaceholder)).toBeInTheDocument();
  });

  it('Should trigger onChange on Password input', async () => {
    const { getByTestId } = render(
      <InputPassword placeholder={inputPlaceholder} onChange={onChangeMock} />,
    );
    const mockChange = 'value';

    expect(getByTestId('input-password')).toBeInTheDocument();

    act(() => {
      fireEvent.change(getByTestId('input-password'), {
        target: { value: mockChange },
      });
    });

    await waitFor(() => {
      expect(onChangeMock).toHaveBeenCalled();
    });
  });
});
