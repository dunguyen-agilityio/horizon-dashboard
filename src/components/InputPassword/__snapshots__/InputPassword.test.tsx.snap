// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputPassword tests Should match snapshot 1`] = `
<div>
  <div
    class="group flex flex-col data-[hidden=true]:hidden w-full group relative justify-end data-[has-label=true]:mt-[calc(theme(fontSize.small)_+_12px)]"
    data-filled="true"
    data-filled-within="true"
    data-has-elements="true"
    data-has-label="true"
    data-required="true"
    data-slot="base"
  >
    <div
      class="h-full flex flex-col"
      data-slot="main-wrapper"
    >
      <div
        class="relative w-full inline-flex tap-highlight-transparent flex-row items-center shadow-sm px-3 gap-3 bg-default-100 data-[hover=true]:bg-default-200 group-data-[focus=true]:bg-default-100 h-12 min-h-12 rounded-large transition-background motion-reduce:transition-none !duration-150 outline-none group-data-[focus-visible=true]:z-10 group-data-[focus-visible=true]:ring-2 group-data-[focus-visible=true]:ring-focus group-data-[focus-visible=true]:ring-offset-2 group-data-[focus-visible=true]:ring-offset-background"
        data-slot="input-wrapper"
        style="cursor: text;"
      >
        <label
          class="absolute pointer-events-none origin-top-left rtl:origin-top-right subpixel-antialiased block text-foreground-500 after:content-['*'] after:text-danger after:ml-0.5 rtl:after:ml-[unset] rtl:after:mr-0.5 will-change-auto !duration-200 !ease-out motion-reduce:transition-none transition-[transform,color,left,opacity] group-data-[filled-within=true]:text-foreground group-data-[filled-within=true]:pointer-events-auto pb-0 z-20 top-1/2 -translate-y-1/2 group-data-[filled-within=true]:left-0 left-3 rtl:left-auto rtl:right-3 text-medium group-data-[filled-within=true]:-translate-y-[calc(100%_+_theme(fontSize.small)/2_+_24px)] pe-2 max-w-full text-ellipsis overflow-hidden"
          data-slot="label"
          for="react-aria-:r0:"
          id="react-aria-:r1:"
        >
          Password
        </label>
        <div
          class="inline-flex w-full items-center h-full box-border"
          data-slot="inner-wrapper"
        >
          <input
            aria-label="inputPassword"
            aria-labelledby="react-aria-:r0: react-aria-:r1:"
            aria-required="true"
            class="w-full font-normal bg-transparent !outline-none placeholder:text-foreground-500 focus-visible:outline-none data-[has-start-content=true]:ps-1.5 data-[has-end-content=true]:pe-1.5 file:cursor-pointer file:bg-transparent file:border-0 autofill:bg-transparent bg-clip-text text-medium group-data-[has-value=true]:text-default-foreground"
            data-has-end-content="true"
            data-slot="input"
            data-testid="input-password"
            id="react-aria-:r0:"
            placeholder="Min. 8 characters"
            type="password"
            value=""
          />
          <button
            aria-label="toggle password visibility"
            class="focus:outline-none"
            type="button"
          >
            <i
              class="fill-secondary w-6 h-6 flex justify-center items-center"
            >
              <svg
                data-testid="eye-slash-filled-icon"
                height="19"
                width="22"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M11 4c2.76 0 5 2.24 5 5 0 .65-.13 1.26-.36 1.83l2.92 2.92c1.51-1.26 2.7-2.89 3.43-4.75-1.73-4.39-6-7.5-11-7.5-1.4 0-2.74.25-3.98.7l2.16 2.16C9.74 4.13 10.35 4 11 4ZM1 1.27l2.28 2.28.46.46A11.804 11.804 0 0 0 0 9c1.73 4.39 6 7.5 11 7.5 1.55 0 3.03-.3 4.38-.84l.42.42L18.73 19 20 17.73 2.27 0 1 1.27ZM6.53 6.8l1.55 1.55C8.03 8.56 8 8.78 8 9c0 1.66 1.34 3 3 3 .22 0 .44-.03.65-.08l1.55 1.55c-.67.33-1.41.53-2.2.53-2.76 0-5-2.24-5-5 0-.79.2-1.53.53-2.2Zm4.31-.78 3.15 3.15.02-.16c0-1.66-1.34-3-3-3l-.17.01Z"
                />
              </svg>
            </i>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
